server/src/main/java/se/sperber/cryson/service/CrysonService.java

```diff
@@ -165,31 +158,65 @@ public Response getEntityById(String entityName, Long id, Set<String> associatio
     List<Object> entities = crysonRepository.findByExample(qualifiedEntityClassName(entityName), exampleEntity, associationsToFetch);
-    return serialize(entities, associationsToFetch);
+    Map<String, Map<Long, Set<Long>>> lazyFieldsValues = loadLazyCollectionFields(entities);
+    return serialize(entities, associationsToFetch, associationsToExclude, lazyFieldsValues);
   }
 
-  public Response getAllEntities(String entityName, Set<String> associationsToFetch) {
+  public Response getAllEntities(String entityName, Set<String> associationsToFetch, Set<String> associationsToExclude) {
     List<Object> entities = crysonRepository.findAll(qualifiedEntityClassName(entityName), associationsToFetch);
-    return serialize(entities, associationsToFetch);
+    Map<String, Map<Long, Set<Long>>> lazyFieldsValues = loadLazyCollectionFields(entities);
+    return serialize(entities, associationsToFetch, associationsToExclude, lazyFieldsValues);
   }
 
-  public Response getEntitiesByNamedQuery(String queryName, MultivaluedMap<String, String> queryParameters, Set<String> associationsToFetch) {
+  public Response getEntitiesByNamedQuery(String queryName, MultivaluedMap<String, String> queryParameters,
+                                          Set<String> associationsToFetch, Set<String> associationsToExclude) {
     List<Object> entities = crysonRepository.findByNamedQuery(queryName, queryParameters);
-    return serialize(entities, associationsToFetch);
+    Map<String, Map<Long, Set<Long>>> lazyFieldsValues = loadLazyCollectionFields(entities);
+    return serialize(entities, associationsToFetch, associationsToExclude, lazyFieldsValues);
   }
 
-  public Response getEntitiesByNamedQueryJson(String queryName, Set<String> associationsToFetch, JsonElement parameters) {
+  public Response getEntitiesByNamedQueryJson(String queryName, Set<String> associationsToFetch,
+                                              Set<String> associationsToExclude, JsonElement parameters) {
     List<Object> entities = crysonRepository.findByNamedQueryJson(queryName, parameters);
-    return serialize(entities, associationsToFetch);
+    Map<String, Map<Long, Set<Long>>> lazyFieldsValues = loadLazyCollectionFields(entities);
+    return serialize(entities, associationsToFetch, associationsToExclude, lazyFieldsValues);
+  }
+
+  private Map<String, Map<Long, Set<Long>>> loadLazyCollectionFields(List<Object> entities) {
+    if (entities.size() == 0) return Collections.emptyMap();
+    final Set<Long> objectIds = entities.stream().map(reflectionHelper::getPrimaryKey).collect(Collectors.toSet());
+    Object entity = entities.get(0);
+    if (entity instanceof LazyCollectionSupport) {
+      Map<String, Map<Long, Set<Long>>> map = new HashMap<>();
+      Map<String, String> queries = ((LazyCollectionSupport) entity).lazyCollectionFieldQueries();
+      for (String field : queries.keySet()) {
```

Comment:
Don't know if it gives much but we could potentially skip a few queries if we also consider the includes and excludes list. As it looks to me, if a field is included or excluded then it will not use the map created here.

Reply:
you are definite right. There are places can be more optimised